src/thunderbots/software/ai/navigator/path_planner/astar.cpp

```diff
@@ -0,0 +1,163 @@
+//
+#include "astar.h"
+
+#include <boost/property_map/function_property_map.hpp>
+#include <boost/unordered_map.hpp>
+
+#include "geom/util.h"
+
+// struct grid_vertex_hash
+std::size_t AStar::grid_vertex_hash::operator()(const GridVertex &gp) const
+{
+    std::size_t seed = 0;
+    boost::hash_combine(seed, gp[0]);
+    boost::hash_combine(seed, gp[1]);
+}
+
+// AStarGridGraph
+AStar::AStarGridGraph::AStarGridGraph(const Field &field)
+    : field_min_x(-(field.length() / 2)), field_min_y(-(field.width() / 2))
+{
+    // TODO: allow configuring the field graph number of nodes with
+    // dynamic parameters
+    size_t field_length_nodes                = field.totalLength() * GRID_VERTEX_DENSITY;
+    size_t field_width_nodes                 = field.totalWidth() * GRID_VERTEX_DENSITY;
+    boost::array<size_t, 2> graph_dimensions = {field_length_nodes, field_width_nodes};
+
+    field_graph = std::make_unique<GridGraph2D>(graph_dimensions);
+
+    GridGraph2D::vertex_iterator v, vbegin, vend;
+    for (boost::tie(v, vend) = vertices(*field_graph); v != vend; v++)
+    {
+        Point p = gridPointToPoint(*v);
+        grid_points_list.emplace_back(p, *v);
+    }
+}
+
+Point AStar::AStarGridGraph::gridPointToPoint(const GridVertex &grid_v)
+{
+    double step_size = 1.0f / GRID_VERTEX_DENSITY;
+    Point p(field_min_x + grid_v[0] * step_size, field_min_y + grid_v[1] * step_size);
+    return p;
+}
+
+AStar::GridVertex AStar::AStarGridGraph::nearestGridVertex(const Point &point)
+{
+    // this is unfortunately O(n) w.r.t. number of vertices
+    // try not to call this too much during the actual search
+    auto closest_point_it = grid_points_list.begin();
+    std::nth_element(grid_points_list.begin(), closest_point_it, grid_points_list.end(),
+                     [point](const auto &a, const auto &b) {
+                         return dist(a.first, point) < dist(b.first, point);
+                     });
+    return closest_point_it->second;
+}
+
+const AStar::GridGraph2D &AStar::AStarGridGraph::graph()
+{
+    return *field_graph;
+}
+
+constexpr double AStar::AStarGridGraph::gridPointDistance()
+{
+    return 1.0f / GRID_VERTEX_DENSITY;
+}
+
+// AStarHeuristic
+AStar::AStarHeuristic::AStarHeuristic(
+    const std::shared_ptr<AStar::AStarGridGraph> &_graph, const Point &_dest)
+    : graph(_graph), dest_point(_dest)
+{
+}
+
+AStar::edge_cost_t AStar::AStarHeuristic::operator()(AStar::GridVertex gp)
+{
+    // TODO: add properly scaled obstacle component to heuristic
+    Point p = graph->gridPointToPoint(gp);
+    return dist(dest_point, p);
+}
+
+// AStarVertexVisitor
+AStar::AStarVertexVisitor::AStarVertexVisitor(AStar::GridVertex _dest) : dest(_dest) {}
+
+void AStar::AStarVertexVisitor::examine_vertex(GridVertex grid_v,
+                                               const GridGraph2D &graph)
+{
+    if (grid_v == dest)
+    {
+        // throw an exception because the people who came up with boost
+        // are crazy
```

Comment:
Ahahaha. This is a really interesting way of doing things..... I guess it makes sense in the sense that once you've found the goal you want to stop everything ASAP, but still...... anyhow, we should maybe put a more professional comment here :stuck_out_tongue:

Reply:
fixed